### PR TITLE
Better #inspect for core classes

### PR DESCRIPTION
**Pull request by zverok**
Main changes:
* all main objects' `inspect`s changed to represent data, not internal state, and be more compact in this;
* factor out logic of "table with adaptive column width and limited rows count" into `Formatters::Table`;
* much better (or I hope so) inspect for multi-indexes, multi-index vectors and dataframes.

Examples:

**Daru::Index**:
```ruby
Daru::Index.new('a'..'z')

# before
# => #<Daru::Index:0x93ac470 @relation_hash={"a"=>0, "b"=>1, "c"=>2, "d"=>3, "e"=>4, "f"=>5, "g"=>6, "h"=>7, "i"=>8, "j"=>9, "k"=>10, "l"=>11, "m"=>12, "n"=>13, "o"=>14, "p"=>15, "q"=>16, "r"=>17, "s"=>18, "t"=>19, "u"=>20, "v"=>21, "w"=>22, "x"=>23, "y"=>24, "z"=>25}, @keys=["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"], @size=26>

# after
# => #<Daru::Index(26): {a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t ... z}>
```

**Daru::MultiIndex**:
```ruby
mi = Daru::MultiIndex.from_tuples [
  [:a,:one,:bar],
  [:a,:one,:baz],
  [:a,:two,:bar],
  [:a,:two,:baz],
  [:b,:one,:bar],
  [:b,:two,:bar],
  [:b,:two,:baz],
  [:b,:one,:foo],
  [:c,:one,:bar],
  [:c,:one,:baz],
  [:c,:two,:foo],
  [:c,:two,:bar]
]

# before
#  => #<Daru::MultiIndex:77179890 (levels: [[:a, :b, :c], [:one, :two], [:bar, :baz, :foo]]
# labels: [[0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2], [0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1], [0, 1, 0, 1, 0, 0, 1, 2, 0, 1, 2, 0]])> 

# after
# => #<Daru::MultiIndex(3x12)>
#   a one bar
#         baz
#     two bar
#         baz
#   b one bar
#     two bar
#         baz
#     one foo
#   c one bar
#         baz
#     two foo
#         bar 
```

**Daru::Vector**:
```ruby
Daru::Vector.new [1,2,3], name: 'foo'
# => #<Daru::Vector:77150670 @name = foo @metadata = {} @size = 3 >
#    foo
#  0   1
#  1   2
#  2   3 

# after
# => #<Daru::Vector(3)>
#     foo
#   0   1
#   1   2
#   2   3

# with multi-index
Daru::Vector.new (1..12), index: mi, name: 'foo'

# before
# => #<Daru::Vector:77029920 @name = foo @metadata = {} @size = 12 >
#                               foo
# [:a, :one, :bar]                1
# [:a, :one, :baz]                2
# [:a, :two, :bar]                3
# [:a, :two, :baz]                4
# [:b, :one, :bar]                5
# [:b, :two, :bar]                6
# [:b, :two, :baz]                7
# [:b, :one, :foo]                8
# [:c, :one, :bar]                9
# [:c, :one, :baz]               10
# [:c, :two, :foo]               11
# [:c, :two, :bar]               12 

# after
# => #<Daru::Vector(12)>
#             foo
#   a one bar   1
#         baz   2
#     two bar   3
#         baz   4
#   b one bar   5
#     two bar   6
#         baz   7
#     one foo   8
#   c one bar   9
#         baz  10
#     two foo  11
#         bar  12 
```

**Daru::DataFrame**:
```ruby
Daru::DataFrame.new({a: [1,2,3], b: [3,4,5], c: [6,7,8]}, name: 'test')

# before
# => #<Daru::DataFrame:76902890 @name = test @size = 3>
#      a   b   c
#  0   1   3   6
#  1   2   4   7
#  2   3   5   8 


# after
# => #<Daru::DataFrame: test (3x3)>
#       a   b   c
#   0   1   3   6
#   1   2   4   7
#   2   3   5   8 

# with multi-index

Daru::DataFrame.new({a: (1..12), b: ('a'..'l'), c: (1001..1012)}, index: mi, name: 'test')

# before
#  => #<Daru::DataFrame:76819850 @name = test @size = 12>
#                     a          b          c
# [:a, :one,          1          a       1001
# [:a, :one,          2          b       1002
# [:a, :two,          3          c       1003
# [:a, :two,          4          d       1004
# [:b, :one,          5          e       1005
# [:b, :two,          6          f       1006
# [:b, :two,          7          g       1007
# [:b, :one,          8          h       1008
# [:c, :one,          9          i       1009
# [:c, :one,         10          j       1010
# [:c, :two,         11          k       1011
# [:c, :two,         12          l       1012 

# after
# => #<Daru::DataFrame: test (3x12)>
#                   a    b    c
#    a  one  bar    1    a 1001
#            baz    2    b 1002
#       two  bar    3    c 1003
#            baz    4    d 1004
#    b  one  bar    5    e 1005
#       two  bar    6    f 1006
#            baz    7    g 1007
#       one  foo    8    h 1008
#    c  one  bar    9    i 1009
#            baz   10    j 1010
#       two  foo   11    k 1011
#            bar   12    l 1012 
```

### CODE REVIEW

File: lib/daru/dataframe.rb [hunk 1]
```diff
@@ -237,7 +237,7 @@ def from_vector_rows source, opts
     def initialize source, opts={} # rubocop:disable Metrics/MethodLength
       vectors, index = opts[:order], opts[:index] # FIXME: just keyword arges after Ruby 2.1
       @data = []
-      @name = opts[:name] || SecureRandom.uuid
+      @name = opts[:name]
```
Comment (zverok) [inline lib/daru/dataframe.rb]: As per https://github.com/v0dro/daru/issues/144

Comment (v0dro) [inline lib/daru/dataframe.rb]: Please put this in a different PR. It needs to show up in the commit history as a separate feature.

**Comment by v0dro** [comment]
Nice!

File: spec/dataframe_spec.rb [hunk 2]
```diff
@@ -2663,69 +2656,112 @@ def create_test(*args, &_proc)
...
-      it { should == %Q{
-        |#<Daru::DataFrame:#{df.object_id} @name = test @size = 3>
-        |                    a          b          c
-        |         0          1          4 this is ri
-        |         1          2          5        nil
-        |         2          3          6        nil
+      it { is_expected.to eq %Q{
+        |#<Daru::DataFrame: test (3x3)>
+        |                     a          b          c
+        |          0          1          4 this is ri
```
Comment (v0dro) [inline spec/dataframe_spec.rb]: Scope for improvement: print this as `this is...`

File: lib/daru/dataframe.rb (hunk 1 above)
Comment (zverok) [inline lib/daru/dataframe.rb]: Ah, ok, sure. https://github.com/v0dro/daru/pull/147

Comment (v0dro) [inline lib/daru/dataframe.rb]: You'll need to revert this from here too :P

File: spec/dataframe_spec.rb (hunk 2 above)
Comment (zverok) [inline spec/dataframe_spec.rb]: Yeah, thought about it, but was a bit lazy, faithfully :)

File: lib/daru/dataframe.rb (hunk 1 above)
Comment (zverok) [inline lib/daru/dataframe.rb]: I'll just wait while you merge https://github.com/v0dro/daru/pull/147, and then rebase from master, and this PR would look OK :)

**Comment by lokeshh** [comment]
> # => #<Daru::DataFrame: test (3x12)>
How about `12x3` instead of `3x12`? `rows x cols` is more popular I think.

**Comment by v0dro** [comment]
Yeah that would be better. It's also used a lot that way for matrix notation. @zverok ?

**Comment by zverok** [comment]
Of course, if it is more common format, definitely should be changed. Will do a separate PR tomorrow.

**Comment by zverok** [comment]
What about `MultiIndex`? I've also did it in this order:

```ruby
# => #<Daru::MultiIndex(3x12)>
#   a one bar
#         baz
#     two bar
#         baz
#   b one bar
#     two bar
#         baz
#     one foo
#   c one bar
#         baz
#     two foo
#         bar 
```

Is it sane? Or should be changed somehow?..

**Comment by v0dro** [comment]
Lets follow the same conventions throughout the repo. (row x col).
